code/modules/reagents/chemistry/machinery/pandemic.dm

```diff
@@ -68,117 +79,223 @@
+				return
+
+			var/obj/item/reagent_containers/glass/bottle/B = create_culture(vaccine_name, "vaccine", 200)
+			B.reagents.add_reagent("vaccine", 15, list(vaccine_type))
+		if("eject_beaker")
+			eject_beaker()
+		if("destroy_eject_beaker")
+			beaker.reagents.clear_reagents()
+			eject_beaker()
+		if("print_release_forms")
+			var/strain_index = text2num(params["strain_index"])
+			if(isnull(strain_index))
+				atom_say("Unable to respond to command.")
+				return
+			var/type = GetVirusTypeByIndex(strain_index)
+			if(!type)
+				atom_say("Unable to find requested strain.")
+				return
+			var/datum/disease/advance/A = GLOB.archive_diseases[type]
+			if(!A)
+				atom_say("Unable to find requested strain.")
+				return
+			print_form(A, usr)
+		if("name_strain")
+			var/strain_index = text2num(params["strain_index"])
+			if(isnull(strain_index))
+				atom_say("Unable to respond to command.")
+				return
+			var/type = GetVirusTypeByIndex(strain_index)
+			if(!type)
+				atom_say("Unable to find requested strain.")
+				return
+			var/datum/disease/advance/A = GLOB.archive_diseases[type]
+			if(!A)
+				atom_say("Unable to find requested strain.")
+				return
+			if(A.name != "Unknown")
+				atom_say("Request rejected. Strain already has a name.")
+				return
+			var/new_name = tgui_input_text(usr, "Name the Strain", "New Name", max_length = MAX_NAME_LEN)
+			if(!new_name)
+				return
 			A.AssignName(new_name)
 			for(var/datum/disease/advance/AD in GLOB.active_diseases)
 				AD.Refresh()
-		updateUsrDialog()
-	else if(href_list["print_form"])
-		var/datum/disease/D = GetVirusByIndex(text2num(href_list["print_form"]))
-		D = GLOB.archive_diseases[D.GetDiseaseID()]//We know it's advanced no need to check
-		print_form(D, usr)
+		if("switch_strain")
+			var/strain_index = text2num(params["strain_index"])
+			if(isnull(strain_index) || strain_index < 1)
+				atom_say("Unable to respond to command.")
+				return
+			var/list/viruses = GetViruses()
+			if(strain_index > length(viruses))
+				atom_say("Unable to find requested strain.")
+				return
+			selected_strain_index = strain_index;
+		else
+			return FALSE
 
+/obj/machinery/computer/pandemic/ui_state(mob/user)
+	return GLOB.default_state
 
-	else
-		usr << browse(null, "window=pandemic")
-		updateUsrDialog()
-		return
+/obj/machinery/computer/pandemic/ui_interact(mob/user, datum/tgui/ui = null)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PanDEMIC", name)
+		ui.open()
 
-	add_fingerprint(usr)
+/obj/machinery/computer/pandemic/ui_data(mob/user)
+	var/datum/reagent/blood/Blood = null
+	if(beaker)
+		var/datum/reagents/R = beaker.reagents
+		for(var/datum/reagent/blood/B in R.reagent_list)
+			if(B)
+				Blood = B
+				break
+
+	var/list/data = list(
+		"synthesisCooldown" = !!wait,
+		"beakerLoaded" = !!beaker,
+		"beakerContainsBlood" = !!Blood,
+		"beakerContainsVirus" = !!length(Blood?.data["viruses"]),
```

Comment:
```suggestion
		"beakerContainsVirus" = !!length(Blood?.data["viruses"]),
```
I personally don't like that style, but anyway
I'm not really sure what's going on here